### PR TITLE
stbt.match: Add a new sqdiff normalisation method

### PR DESCRIPTION
**Pull request by drothlis**
OpenCV's sqdiff-normed normalises according to the pixel intensities (see [cv2.matchTemplate]). We can use sqdiff instead and then do our own normalisation (based solely on the reference-image size).

This is needed for #542 because OpenCV's templatematch support for transparency only works with sqdiff (not sqdiff-normed).

[cv2.matchTemplate]: https://docs.opencv.org/2.4/modules/imgproc/doc/object_detection.html#cv2.matchTemplate

TODO:

- [x] Check ranges of the value from OpenCV's sqdiff-normed.
- [x] Enums for existing MatchMethod & ConfirmMethod values.
- [x] New SQDIFF normalisation.
- [x] Tests.

Future TODOs:

- Make it the default.
    - Figure out appropriate threshold (initial results suggest it can be much higher, like .95)

### CODE REVIEW

**Comment by drothlis** [comment]
> Check ranges of the value from OpenCV's sqdiff-normed.

Yes, OpenCV's SQDIFF_NORMED normalises by the pixel intensity across the reference image and the source image patch. This doesn't work at all for completely black images (division by zero), and it exaggerates differences for very dark images.

See this table showing the first-pass result for 10x10 source & references images, each of uniform intensity. Absolute differences of 30 per pixel still match at >98% for almost-white images, but don't match for almost-black images.

```
      000      001      010      020      030      100      225      235      245      254      255
000:  0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 
001:  0.000000 1.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 
010:  0.000000 0.000000 1.000000 0.500000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 
020:  0.000000 0.000000 0.500000 1.000000 0.833333 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 
030:  0.000000 0.000000 0.000000 0.833333 1.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 
100:  0.000000 0.000000 0.000000 0.000000 0.000000 1.000000 0.305556 0.224468 0.141837 0.066299 0.057843 
225:  0.000000 0.000000 0.000000 0.000000 0.000000 0.305556 1.000000 0.998109 0.992744 0.985284 0.984314 
235:  0.000000 0.000000 0.000000 0.000000 0.000000 0.224468 0.998109 1.000000 0.998263 0.993952 0.993325 
245:  0.000000 0.000000 0.000000 0.000000 0.000000 0.141837 0.992744 0.998263 1.000000 0.998698 0.998399 
254:  0.000000 0.000000 0.000000 0.000000 0.000000 0.066299 0.985284 0.993952 0.998698 1.000000 0.999984 
255:  0.000000 0.000000 0.000000 0.000000 0.000000 0.057843 0.984314 0.993325 0.998399 0.999984 1.000000 
```

Generated with:

```
import numpy, stbt

intensities = [0, 1, 10, 20, 30, 100, 225, 235, 245, 254, 255]

print "      " + "      ".join("%03d" % i for i in intensities)
for a in intensities:
    print "%03d: " % a,
    source = numpy.ones((10,10,3), numpy.uint8) * a
    for b in intensities:
        reference = numpy.ones((10,10,3), numpy.uint8) * b
        result = stbt.match(
            reference, source,
            stbt.MatchParameters(match_method="sqdiff-normed"))
        print "%.6f" % (result.first_pass_result,),
    print ""
```

**Comment by drothlis** [comment]
For comparison here are the first-pass results with `MatchMethod.SQDIFF` (doing our own normalisation):

```
      000      001      010      020      030      100      225      235      245      254      255
000:  1.000000 0.999985 0.998462 0.993849 0.986159 0.846213 0.221453 0.150711 0.076894 0.007828 0.000000 
001:  0.999985 1.000000 0.998754 0.994448 0.987067 0.849273 0.228358 0.157924 0.084414 0.015625 0.007828 
010:  0.998462 0.998754 1.000000 0.998462 0.993849 0.875433 0.289120 0.221453 0.150711 0.084414 0.076894 
020:  0.993849 0.994448 0.998462 1.000000 0.998462 0.901576 0.353710 0.289120 0.221453 0.157924 0.150711 
030:  0.986159 0.987067 0.993849 0.998462 1.000000 0.924644 0.415225 0.353710 0.289120 0.228358 0.221453 
100:  0.846213 0.849273 0.875433 0.901576 0.924644 1.000000 0.759708 0.719723 0.676663 0.635279 0.630527 
225:  0.221453 0.228358 0.289120 0.353710 0.415225 0.759708 1.000000 0.998462 0.993848 0.987067 0.986159 
235:  0.150711 0.157924 0.221453 0.289120 0.353710 0.719723 0.998462 1.000000 0.998462 0.994448 0.993848 
245:  0.076894 0.084414 0.150711 0.221453 0.289120 0.676663 0.993848 0.998462 1.000000 0.998754 0.998462 
254:  0.007828 0.015625 0.084414 0.157924 0.228358 0.635279 0.987067 0.994448 0.998754 1.000000 0.999984 
255:  0.000000 0.007828 0.076894 0.150711 0.221453 0.630527 0.986159 0.993848 0.998462 0.999984 1.000000 
```

**Comment by drothlis** [comment]
@wmanley Please review. This adds the new MatchMethod but doesn't make it the default. I've raised #544 for that, so that we have more time to decide if it's a good idea.

File: _stbt/match.py [hunk 1]
```diff
@@ -598,10 +599,13 @@ def _match_template(image, template, method, roi_mask, level, imwrite):
...
 
     matches_heatmap = (
-        (numpy.ones if method == cv2.TM_SQDIFF_NORMED else numpy.zeros)(
+        (numpy.ones if method in (cv2.TM_SQDIFF, cv2.TM_SQDIFF_NORMED)
+         else numpy.zeros)(
             (image.shape[0] - template.shape[0] + 1,
              image.shape[1] - template.shape[1] + 1),
             dtype=numpy.float32))
+    if method == cv2.TM_SQDIFF:
+        matches_heatmap *= template.size * (255 ** 2)
```
[the remark this answers is not in the document]
Comment (wmanley) [reply _stbt/match.py]:
```
    heatmap_shape = (
            (image.shape[0] - template.shape[0] + 1,
             image.shape[1] - template.shape[1] + 1)
    BAD_VALUE = {
        cv2.TM_SQDIFF_NORMED: 1,
        cv2.TM_CCOEFF_NORMED: 0,
        cv2.TM_CCORR_NORMED: 0,
        cv2.TM_SQDIFF: template.size * (255 ** 2)
    }
    matches_heatmap = numpy.full(heatmap_shape, BAD_VALUE[method], dtype=numpy.float32)
```

File: tests/test_match.py [hunk 2]
```diff
@@ -110,23 +143,51 @@ def test_that_results_dont_overlap():
...
     print matches
     assert matches == [stbt.Region(x, y, width=135, height=44) for x, y in [
         (177, 75), (177, 119)]]
 
 
-def test_match_all_with_an_image_that_matches_everywhere():
+@pytest.mark.parametrize("match_method", [
+    stbt.MatchMethod.SQDIFF,
+    stbt.MatchMethod.SQDIFF_NORMED,
+])
```
Comment (wmanley) [inline tests/test_match.py]: I like this test parameterisation.

File: _stbt/config.py [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import ConfigParser
+import enum
```
[the remark this answers is not in the document]
Comment (drothlis) [reply _stbt/config.py]: > enum is not a part of the standard library so this should be lower.

It needs to be there for pylint and for future python 3 compatibility.

> We need to consider whether non-stdlib dependencies are appropriate in this file too.

It's a backport from the Python 3 stdlib, and it doesn't have any dependencies other than Python itself. I think it's OK.

> I can't see why we'd want to allow `2` in this instance?

Allowing `"sqdiff-normed"` (for `MatchMethod`) is the real reason.

File: _stbt/match.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (drothlis) [reply _stbt/match.py]: Done, thanks.

**Comment by drothlis** [comment]
> I think sqdiff could be the first-pass method we've always dreamed about.

Yes, it seems to work much better. I never realised that OpenCV's normalisation was based on the pixel intensities rather than the number of pixels. Thanks @wmanley for providing the idea!

**Pull request merged**